go/vt/discovery/keyspace_events.go

```diff
@@ -329,6 +377,96 @@ func (kss *keyspaceState) onHealthCheck(th *TabletHealth) {
 	kss.ensureConsistentLocked()
 }
 
+type MoveTablesType int
+
+const (
+	MoveTablesNone MoveTablesType = iota
+	MoveTablesRegular
+	MoveTablesShardByShard
+)
+
+type MoveTablesState struct {
+	Typ   MoveTablesType
+	State MoveTablesStatus
+}
+
+func (kss *keyspaceState) getMoveTablesStatus(vs *vschemapb.SrvVSchema) (*MoveTablesState, error) {
+	mtState := &MoveTablesState{
+		Typ:   MoveTablesNone,
+		State: MoveTablesUnknown,
+	}
+
+	// if there are no routing rules defined, then movetables is not in progress, exit early
+	if (vs.RoutingRules != nil && len(vs.RoutingRules.Rules) == 0) &&
+		(vs.ShardRoutingRules != nil && len(vs.ShardRoutingRules.Rules) == 0) {
+		return mtState, nil
+	}
+
+	ctx := kss.kew.ctx
+	ts, _ := kss.kew.ts.GetTopoServer()
+
+	// collect all current shard information from the topo
+	var shardInfos []*topo.ShardInfo
+	for _, sstate := range kss.shards {
+		si, err := ts.GetShard(ctx, kss.keyspace, sstate.target.Shard)
+		if err != nil {
+			return nil, err
+		}
+		shardInfos = append(shardInfos, si)
+	}
+
+	// check if any shard has denied tables and if so, record one of these to check where it currently points to
+	// using the (shard) routing rules
+	var shardsWithDeniedTables []string
+	var oneDeniedTable string
+	for _, si := range shardInfos {
+		for _, tc := range si.TabletControls {
+			if len(tc.DeniedTables) > 0 {
+				oneDeniedTable = tc.DeniedTables[0]
+				shardsWithDeniedTables = append(shardsWithDeniedTables, si.ShardName())
+			}
+		}
+	}
+	if len(shardsWithDeniedTables) == 0 {
+		return mtState, nil
+	}
+
+	// check if a shard by shard migration is in progress and if so detect if it has been switched
+	isPartialTables := vs.ShardRoutingRules != nil && len(vs.ShardRoutingRules.Rules) > 0
+
+	if isPartialTables {
+		srr := topotools.GetShardRoutingRulesMap(vs.ShardRoutingRules)
+		mtState.Typ = MoveTablesShardByShard
+		mtState.State = MoveTablesSwitched
+		for _, shard := range shardsWithDeniedTables {
+			ruleKey := topotools.GetShardRoutingRuleKey(kss.keyspace, shard)
+			if _, ok := srr[ruleKey]; ok {
+				// still pointing to the source shard
+				mtState.State = MoveTablesSwitching
+				break
+			}
+		}
+		log.Infof("getMoveTablesStatus: keyspace %s declaring partial move tables %v", kss.keyspace, mtState)
+		return mtState, nil
+	}
+
+	// it wasn't a shard by shard migration, but since we have denied tables it must be a regular MoveTables
+	mtState.Typ = MoveTablesRegular
+	mtState.State = MoveTablesSwitching
+	rr := topotools.GetRoutingRulesMap(vs.RoutingRules)
+	if rr != nil {
+		r, ok := rr[oneDeniedTable]
+		// if a rule exists for the table and points to the target keyspace, writes have been switched
+		if ok && len(r) > 0 && r[0] != fmt.Sprintf("%s.%s", kss.keyspace, oneDeniedTable) {
+			mtState.State = MoveTablesSwitched
+			log.Infof("onSrvKeyspace::  keyspace %s writes have been switched for table %s, rule %v", kss.keyspace, oneDeniedTable, r[0])
```

Comment:
as discussed offline, we should use vindexes.VSchema struct and the helper methods available in it.